### ISSUE TITLE
Resetting state variables

### ISSUE THREAD

**Issue by spinang** [genn-team/genn#105]
Good morning,
in one of my previous post I asked you how I can reset a given neurons population restoring the initial conditions that I specify in the .cc file during the execution of .cu file.

You explained me how to proceed in that post but I've not understand very well, so I ask you gently if you can explain me with a practical example, assuming for instance a population P with initial conditions I[2] = {a,b} (a and b are two generic numbers).

Moreover, another question I would ask you is the following: let us suppose we have a text file in which I save some initial conditions. If I want to upload them during the execution of .cu file and to store them in my neuron (or synapse) group, is such procedure similar to the one related to my first question?

Thanks in advance.

**Comment by jamesturner246** [comment]
For example, to set variable `V` of neuron group `MyGroup` to initial conditions, you would store the initial conditions inside a host array - say `initialMyGroupV` - in user code. You would copy that array into the host-side version of the array for that group / variable in user code, and then copy the host-side arrays to device side using the `copyMyGroupStateToDevice` function generated by GeNN:

```C++
double initialMyGroupV[5] = { -80.0, -40.1, -78.0, -60.0, -81.2};

// copy initial conditions into host-side array
std::copy(initialMyGroupV, initialMyGroupV + 5, MyGroupV);

// copy the host-side arrays for MyGroup onto the device
copyMyGroupStateToDevice();
```

To load the initial conditions from file, you would use the `ifstream` C++ class:

```C++
ifstream myFile;

// open a file with initial values separated by spaces
myFile.open("file_with_initial_conditions.txt");

// read initial values into host-side array
for (int i = 0; i < 5; i++) {
    myFile >> MyGroupV[i];
}

// copy the host-side arrays for MyGroup onto the device
copyMyGroupStateToDevice();

// close the initial values file
myFile.close();
```

So basically you just overwrite the host-side variable arrays with your initial values, and then copy those arrays onto device with the functions generated by GeNN, inside user code. By user code I mean the code which copies data to device and calls `timeStepGPU`, etcetera. 

All the best,
James.

**Comment by jamesturner246** [comment]
NB: Obviously the functions used for C are different, since C doesn't have `std::ifstream` and `std::copy`, but the principle is the same.

**Comment by spinang** [comment]
Thank you for you reply, I will take your advice and see if it works for my application.

**Comment by spinang** [comment]
Does `copyMyGroupStateToDevice()` require any additional libraries to be included?
Because after trying to compiling my code, the systems tells me the function is undefined. Same situation occurs if I try to do something like these:
`char * neuronName = NNmodel.neuronName[neuronGroup];`
In this second case, the error I get is `type name is not allowed`.

**Comment by spinang** [comment]
Up, please.

**Comment by jamesturner246** [comment]
Hi spinang. Sorry; this one slipped my attention.

The first problem was my mistake. I told you the function is called `copyMyGroupStateToDevice`, but it's actually called `pushInputStateToDevice`. It is declared in the header 'MyModel_CODE/definitions.h', and defined in 'MyModel_CODE/runnerGPU.c' after the code-generation phase. Take a look for yourself to see what other functions are generated for the user's code.

The second problem is likely due to the fact that you are using a class name `NNmodel` as a variable name. Change it to `model`, or something. Furthermore, you are assigning a string to a char array without doing any conversion. Try changing `neuronName` to a string, or using a C++ standard library function to convert `model.neuronName[neuronGroup]` into a char array.

**Comment by jamesturner246** [comment]
Note the edit. The function, in general, is `push<GroupName>StateToDevice`.

Best,
James..

**Comment by spinang** [comment]
This is a test model:

```
#define DT 0.1 
#include "modelSpec.h"
#include "modelSpec.cc"
void modelDefinition(NNmodel &model)
{
	// definition of tenHHModel
	initGeNN();
	model.setName("RESET");
	double p[7] = {
		7.15,          // 0 - gNa: Na conductance in muS
		50.0,          // 1 - ENa: Na equi potential in mV
		1.43,          // 2 - gK: K conductance in muS
		-95.0,          // 3 - EK: K equi potential in mV 
		0.02672,       // 4 - gl: leak conductance in muS
		-63.563,       // 5 - El: leak equi potential in mV
		0.143          // 6 - Cmem: membr. capacity density in nF
	};
	double ini[4] = {
		-60.0,         // 0 - membrane potential V
		0.0529324,     // 1 - prob. for Na channel activation m
		0.3176767,     // 2 - prob. for not Na channel blocking h
		0.5961207      // 3 - prob. for K channel activation n
	};
	model.addNeuronPopulation("Pop1", 1, TRAUBMILES, p, ini);

	model.finalize();
}
```
and this is the simulation file:
```
#include "c.cc"
#include "RESET_CODE/runner.cc"
#include "RESET_CODE/definitions.h"
#include "C:\genn-team-genn-8cacb78\lib\include\hr_time.cpp"
#include <algorithm>

void reset();

int main()
{
	const double tReset = 30.0;
	CStopWatch timer;
	timer.startTimer();
	allocateMem();
	initialize();
	ofstream os("output.V.dat");
		
	for(double t = 0.0; t < 50; t+=DT){
		if(t == tReset) reset();
		stepTimeGPU(t);
		pullPop1StateFromDevice();
		os << t << "\t" << VPop1[0] << endl;
	}
	
	os.close();
	timer.stopTimer();
	cout << "The simulation time is: " << timer.getElapsedTime() << " seconds " << endl;
	return 0;
}

void reset(){
	double ini[4] = {
		-60.0,         // 0 - membrane potential V
		0.0529324,     // 1 - prob. for Na channel activation m
		0.3176767,     // 2 - prob. for not Na channel blocking h
		0.5961207      // 3 - prob. for K channel activation n
	};
	
	std::copy(ini, ini + 4, Pop1);
	pushPop1StateToDevice();
}
```
Of course, Pop1 is not recognized as the right parameter! But is it right replacing it with VPop1? It seems it doesn't correspond to with I'd to do... or more probably I'm doing it wrong.

NB: I've tried to write VPop1 instead of Pop1, but it doesn't work. What I get is:

```
Microsoft (R) Program Maintenance Utility Version 12.00.21005.1
Copyright (C) Microsoft Corporation.  All rights reserved.

        del s *.obj *.ilk *.pdb 2>nul

Microsoft (R) Program Maintenance Utility Version 12.00.21005.1
Copyright (C) Microsoft Corporation.  All rights reserved.

        "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.0\bin\nvcc.exe"  -g -G
 -arch sm_21 -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.0\include" -I"C
:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.0\samples\common\inc" -I"C:\genn
-team-genn-8cacb78\lib\include" -I"C:\genn-team-genn-8cacb78\userproject\include"   s
.cu /Fos.obj -c
s.cu
C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xutility(2045) : warni
ng C4244: '=' : conversion from 'double' to 'float', possible loss of data
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xutility(2066)
 : see reference to function template instantiation '_OutIt std::_Copy_impl<_InIt,_Ou
tIt>(_InIt,_InIt,_OutIt,std::_Nonscalar_ptr_iterator_tag)' being compiled
        with
        [
            _OutIt=float *
,            _InIt=double *
        ]
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\xutility(2077)
 : see reference to function template instantiation '_OutIt std::_Copy_impl<_Iter,flo
at>(_InIt,_InIt,_OutIt)' being compiled
        with
        [
            _OutIt=float *
,            _Iter=double *
,            _InIt=double *
        ]
        s.cu(39) : see reference to function template instantiation '_OutIt std::copy
<double*,float*>(_InIt,_InIt,_OutIt)' being compiled
        with
        [
            _OutIt=float *
,            _InIt=double *
        ]
Fos.obj
        cl  /nologo /EHsc /Zi "C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v7.
0\lib\x64\cudart.lib" s.obj /Fes
```

**Comment by tnowotny** [comment]
Hi - it took me a moment to understand why this is confusing but I do now. At model definition, one can pass in the initial variables as an array and they are set for the different variables. Subsequently, each variable is its own array. So, in your case, it would take something like
```VPop1[0]= ini[0];
mPop1[0]= ini[1];
hPop1[0]= ini[2];
nPop1[0]= ini[3];
pushPop1StateToDevice();```
It looks a little odd because you have only one neuron in the population (hence just zeros on the lhs expressions, for the 0th neuron).
Hope this helps.

**Comment by tnowotny** [comment]
So what is the result of calling reset()?
I can't find anything wrong with the code on first sight but I also don't know what I am looking for.
A simple test would be to add: 
after line 327:
```
cerr << index << ":" << VFEAUTURE1[index] << endl;
```
then, 
after line 344:
```
pullFEATURE1StateFromDevice();
for (int i= 0; i < NEURONS_PER_COLUMN; i++) {
   cerr << i << ":" << VFEATURE1[i] << endl;
}
```
You should see that everything has the value of init[0], i.e. -65.0
If this is the case, your problems are somewhere else and not in the reset code. 
If not, please let us know what outputs you observed on cerr.

**Comment by spinang** [comment]
Ok, I would to show one of the several files I created for doing some plots in MATLAB.
[THIS FILE](http://ge.tt/1fICLuf2) contains all the membrane voltages for each neuron of each population, as functions of time (first column represents the current time step). Moreover, due to the fact I want to apply several patterns, I used an index to distinguish them (second column), while starting from the third column I save the membrane potentials.
As you can read and accordingly what it has been written before, I expect a reset operation when at least the first iteration ends (then all the neurons should have own V variable equal to -65 (more or less)), but it doesn't occur.

**Comment by tnowotny** [comment]
sorry for delay - I was traveling Friday and today. I can't download your file - something wrong with the sharing site. But what happened when you did the test I asked about?

**Comment by spinang** [comment]
Don't worry.
I fixed the link from which you can download the file ([LINK](http://www.filedropper.com/v_2)). It has been obtained after implementing the instructions for the reset procedure as you told me in the previous post; they should be right, but if you read the attached file you can see that when the pattern changes (for example, passing from the index 0 to the index 1) I expect that membrane potentials re-start again from -65 (or whatever I prefer), but this does not happen.

**Comment by tnowotny** [comment]
I am sorry, I don't know what you are doing. The file you linked has zeros in the first column and seems to have the time from 0 to 2000 ms in the second column. The remainder appears to have nice spiking/bursting Izhikevich neurons.
Please do the test I asked you about and if that fails, let me know. Otherwise, from what you say, it seems that your input isn't changing at all?!?

**Comment by spinang** [comment]
Probably I've understood what's wrong.
Your procedure is effectively correct, because the initial value of V is what I set (more or less), but what makes me confused is what happens _after_ the initial time step. Probably the synaptic connections make the results very different from what I would like to obtain.

For example, [THIS IMAGE](https://postimg.org/image/f5yot2dy5/) shows the time trend of V for a generic neuron during the first and the second simulation (blue line and red line, respectively). Initial conditions are ok, but what I would obtain is that even for the second simulation the trend is similar to the one obtained during the first simulation.

I apologise for wasting your time trying to solve a problem that didn't exist, I'm tremendously sorry.
At the beginning I thought there was an error in my MATLAB script for plotting all the diagrams, but "unfortunately" it is correct too.
